### PR TITLE
Get command uses print-column extn from Openapi schema

### LINKED ISSUE #22: kubectl Get command should use open-api extension to display a resource
kubectl Get command does not provide a rich experience for resources retrieved through federated apiservers and types not compiled into the kubectl binary.

Open-api schema for resources retrieved through federated api servers will have additional metadata as  x-kubernetes-print-columns extension. Kubectl should make use of the extension information to display richer output for such types. So behavior is: if user has not specified any custom output format and x-kubernetes-print-column extension is defined for that type, use extension information to format the output for that resource.

This is part of the detailed proposal described [here](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/sig-cli/get-describe-apiserver-extensions.md) 

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

### PR DESCRIPTION
**Pull request by droot**
**What this PR does / why we need it**:

Kubectl Get command now uses metadata 'x-kubernetes-print-column' from Openapi schema to display a resource. This is to enable richer experience for non-compiled types (like service catalog API resources) in Kubectl. This functionality is currently guarded by a boolean flag "use-openapi-print-columns". 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fixes https://github.com/kubernetes/kubectl/issues/22

**Special notes for your reviewer**:

**Release note**:

```release-note
Get command uses OpenAPI schema to enhance display for a resource if run with flag 'use-openapi-print-columns'. 
An example command:
kubectl get pods --use-openapi-print-columns 
```

### CODE REVIEW

File: pkg/api/v1/types.go
Comment (pwittrock) [inline pkg/api/v1/types.go]: Good for testing, but we don't want to merge this.

File: pkg/kubectl/cmd/get.go
Comment (pwittrock) [inline pkg/kubectl/cmd/get.go]: lets flag guard this, default off until Antoine's optimization is merged

Comment (pwittrock) [inline pkg/kubectl/cmd/get.go]: What  happens to this if the command exits before this is finished?  Should we defer waiting for this to finish or something?

File: pkg/kubectl/cmd/util/factory_builder.go
Comment (pwittrock) [inline pkg/kubectl/cmd/util/factory_builder.go]: should this line be wrapped into: lookupDisplayColumnsFromOpenAPI

File: pkg/kubectl/cmd/util/printing.go
Comment (pwittrock) [inline pkg/kubectl/cmd/util/printing.go]: much better.  add some documentation to the function before merge plz

File: pkg/printers/customcolumn.go
Comment (pwittrock) [inline pkg/printers/customcolumn.go]: what is this and why is it needed?  (add comment)

**Comment by pwittrock** [comment]
generally looks good.  will want to add some tests.

**Review by droot — COMMENTED** [review]
Thanks for the review @pwittrock . Addressed the comments. PTAL.

File: pkg/kubectl/cmd/get.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/get.go]: We should resolve this TODO or at least make it more specific before merging.

File: pkg/kubectl/cmd/get.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/get.go]: This should be returning `(opts *OutputOptions, ok bool)` in accordance with Golang style guides.

File: pkg/printers/interface.go
Comment (alexandercampbell) [inline pkg/printers/interface.go]: Maybe this comment should point where I can find the actual list of possible values.

File: pkg/kubectl/cmd/get.go
Comment (apelisse) [inline pkg/kubectl/cmd/get.go]: Where does the result of this function go? Why is this running asynchronously? How do we not care about the fetched resource?

File: pkg/kubectl/cmd/util/printing.go
Comment (mengqiy) [inline pkg/kubectl/cmd/util/printing.go]: I think we can simplify the signature of `GetStandardPrinter()` by passing in `outputOptions` instead of `outputFormat, templateFile,allowMissingTemplateKeys`.

File: pkg/kubectl/cmd/get.go
Comment (mengqiy) [inline pkg/kubectl/cmd/get.go]: I think we should use `decodedObj`, original may be an unversioned object.
Is there any issue when using `decodedObj`?

File: pkg/kubectl/cmd/util/printing.go
Comment (droot) [inline pkg/kubectl/cmd/util/printing.go]: Done. Thanks for the suggestion.

File: pkg/printers/interface.go
Comment (droot) [inline pkg/printers/interface.go]: I agree. Unfortunately, these are not defined as constant types in the code. For now, I have mentioned the filename where the supported format types are listed inside a function. I would like to make a separate change for defining these valid format types as constants.

**Review by droot — COMMENTED** [review]
Addressed comments and also updated the PR. PTAL.

File: pkg/kubectl/cmd/get.go
Comment (mengqiy) [inline pkg/kubectl/cmd/get.go]: It at least deserves a comment.

**Review by droot — COMMENTED** [review]
Addressed comments and updated the PR with additional comments. @pwittrock @mengqiy PTAL.

**Comment by mengqiy** [comment]
@k8s-bot ok to test

File: pkg/kubectl/cmd/get.go
Comment (pwittrock) [inline pkg/kubectl/cmd/get.go]: why do we need to check Lookup?

**Comment by pwittrock** [comment]
Looks generally correct, but not that familiar with generic printers

File: pkg/kubectl/cmd/get.go
Comment (mengqiy) [inline pkg/kubectl/cmd/get.go]: >--allow-missing-template-keys=true: If true, ignore any errors in templates when a field or map key is missing in the template. Only applies to golang and jsonpath output formats.

I think we can remove the comment.

File: pkg/kubectl/cmd/util/openapi/openapi.go
Comment (mengqiy) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: Why change this?

File: pkg/kubectl/cmd/util/printing.go
Comment (mengqiy) [inline pkg/kubectl/cmd/util/printing.go]: Is `Lookup` necessary?
I think if we have flag `output`, lookup it should always be non-nil, even unspecified.

File: pkg/kubectl/cmd/util/printing.go
Comment (mengqiy) [inline pkg/kubectl/cmd/util/printing.go]: Same question here.

File: pkg/kubectl/cmd/get.go
Comment (droot) [inline pkg/kubectl/cmd/get.go]: cmdutil.GetFlagString(...) does a FATAL if the given flag is not defined. "output" flag is defined for the Get command, so its a bit more defensive check in this case. So can be removed, but doesn't hurt also.

File: pkg/kubectl/cmd/util/printing.go
Comment (droot) [inline pkg/kubectl/cmd/util/printing.go]: Yes, I was being more defensive here, so taking it out.

File: pkg/kubectl/cmd/util/openapi/openapi.go
Comment (droot) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: During testing I discovered for some types, GVK metadata is not defined, so emit this error on stdout which is bad. So didn't have an option but to bump it to higher verbosity.

File: pkg/kubectl/cmd/util/printing.go
Comment (droot) [inline pkg/kubectl/cmd/util/printing.go]: I did a refactor of this code, so kept this code as is to avoid any surprises :)

File: pkg/kubectl/cmd/get.go
Comment (droot) [inline pkg/kubectl/cmd/get.go]: Done. PTAL

**Review by droot — COMMENTED** [review]
Thanks for the review @pwittrock @mengqiy . Addressed the comments. PTAL.

**Comment by droot** [comment]
@k8s-bot ok to test

**Comment by droot** [comment]
@k8s-bot pull-kubernetes-verify test this

**Comment by fabianofranz** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by droot** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by droot** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by droot** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by smarterclayton** [comment]
Um, given how this overlaps with server side get, I'm not sure I approve this.

**Comment by smarterclayton** [comment]
I do not want to add this to the client in 1.7.  I want this to be on the server side.  If we add this to the client, we'll just remove it in 1.8.  What's the point of that?

**Comment by smarterclayton** [comment]
How did this get into the queue without the discussion resolution from the original proposal, saying that this is moving to the server.

**Comment by smarterclayton** [comment]
My primary concern is that this is moving to the server in 1.8, so this flag would be removed and no one would every specify it.  So given that, if the desire to get something in for service catalog is very high, this needs to be an `--experimental-*` flag, and we need to say that the flag will be removed in 1.8.

**Comment by droot** [comment]
@smarterclayton I have updated the PR and renamed the flag to '--experimental-use-openapi-print-columns'. Can you pl. take a look and remove "do-not-merge" label. Lets re-evaluate if we want to deprecate or remove this functionality from kubectl when we move this logic to server side in 1.8.

**Comment by pwittrock** [comment]
> How did this get into the queue without the discussion resolution from the original proposal, saying that this is moving to the server.

@smarterclayton Please refer to the email thread "get consensus on displaying resources - sig-apimachinery / sig-cli" (sent April 17).  There you requested that we get together quickly over VC and resolve the discussion since it had been languishing.  The consensus was that while there was some overlap in the 2 solutions, they were not mutually exclusive.

One of the artifacts from that discussion was - should clients be able to print an object they have without making a request to the server.  e.g. does a Java client that wants to print an object it already fetch need to make a second request to the server to display the object?  Also how can clients display object configuration for resources defined locally or deleted from the server?  We agreed that it would be useful for clients to have some notion for how to print resources without a server roundtrip.

**Comment by pwittrock** [comment]
Making the flag experimental is good idea and we should do that.  I would also make it deprecated so it is hidden.

**Comment by smarterclayton** [comment]
Ok, I vaguely recall us saying:

1. this would be best effort (which means the openapi spec doesn't fit under our normal rules of backwards compatibility)
2. we'd fail gracefully if we can't fetch the columns

I had assumed that when the rest of the server side get plumbing went in, there would be no need for the flag when offline (since we would simply determine that the object was unrecognized, make a best effort pass to get the open api spec, and then print).

The rename to "experimental" satisfies my concerns about the public API shape, objection withdrawn with that change.  For 1.8, I'd expect for us to remove the flag and do the best effort fetch alongside the server side fetch attempt.

**Comment by droot** [comment]
Thanks @smarterclayton 
@pwittrock marked the flag deprecated as well. PTAL. I need LGTM to get this PR merged now.

**Comment by droot** [comment]
@k8s-bot pull-kubernetes-bazel test this

**Comment by pwittrock** [comment]
@smarterclayton SGTM.  Yes, this will be a best effort fetch and very small and fast in 1.8 (when using the zipped binary proto + cache).  Thanks.

**Comment by smarterclayton** [comment]
Why did we introduce OutputOptions instead of adding those fields to PrintOptions?  I don't see anything special about them that justifies a separate object.

If there's no reason to have it separate, I'm going to move it into print options.

**Comment by droot** [comment]
@smarterclayton I think the main distinction is that OutputOptions is used to determine a printer type to use while PrintOptions are simply args to a given Printer to format the output.

**Comment by vencelova** [comment]
IMG_20211101_183930

**Comment by vencelova** [comment]
Haiti

**Comment by vencelova** [comment]
No máme

**Review by hashim21223445 — APPROVED** [review]

**Pull request merged**
